src/xenia/apu/audio_decoder.cpp

```diff
@@ -0,0 +1,239 @@
+
+AudioDecoder::~AudioDecoder() {
+  if (context_) {
+    if (context_->extradata) {
+      delete context_->extradata;
+    }
+    if (avcodec_is_open(context_)) {
+      avcodec_close(context_);
+    }
+    av_free(context_);
+  }
+  if (decoded_frame_) {
+    avcodec_free_frame(&decoded_frame_);
+  }
+  if (current_frame_) {
+    delete current_frame_;
+  }
+}
+
+int AudioDecoder::Initialize(int bits) {
+  static bool avcodec_initialized = false;
+  if (!avcodec_initialized) {
+    avcodec_register_all();
+    avcodec_initialized = true;
+  }
+
+  if (bits <= 0 || bits > 32 || (bits % 8) != 0) {
+    assert_always();
+  }
+
+  // Output bits per sample
+  bits_ = bits;
+
+  // Allocate important stuff
+  codec_ = avcodec_find_decoder(AV_CODEC_ID_WMAPRO);
+  if (!codec_) {
+    return 1;
+  }
+
+  context_ = avcodec_alloc_context3(codec_);
+  if (!context_) {
+    return 1;
+  }
+
+  decoded_frame_ = av_frame_alloc();
+  if (!decoded_frame_) {
+    return 1;
+  }
+
+  packet_ = new AVPacket();
+  av_init_packet(packet_);
+
+  // Initialize these to 0. They'll actually be set later.
+  context_->channels = 0;
+  context_->sample_rate = 0;
+  context_->block_align = XMAContextData::kBytesPerBlock;
+
+  // Extra data passed to the decoder
+  context_->extradata_size = 18;
+  context_->extradata = new uint8_t[18];
+
+  // Current frame stuff whatever
+  // samples per frame * 2 max channels * output bytes
+  current_frame_ = new uint8_t[XMAContextData::kSamplesPerFrame * 2 * (bits / 8)];
+  current_frame_pos_ = 0;
+  frame_samples_size_ = 0;
+
+  *(short *)(context_->extradata) = 0x10; // bits per sample
+  *(int *)(context_->extradata + 2) = 1; // channel mask
+  *(short *)(context_->extradata + 14) = 0x10D6; // decode flags
+
+  // FYI: We're purposely not opening the context here. That is done later.
+
+  return 0;
+}
+
+int AudioDecoder::PreparePacket(uint8_t* input, size_t size,
+                                int sample_rate, int channels) {
+  if (size != XMAContextData::kBytesPerBlock) {
+    // Invalid packet size!
+    assert_always();
+    return 1;
+  }
+  if (packet_->size > 0 || current_frame_pos_ != frame_samples_size_) {
+    // Haven't finished parsing another packet
+    return 1;
+  }
+
+  std::memcpy(packet_data_, input, size);
+
+  // Modify the packet header so it's WMAPro compatible
+  *((int *)packet_data_) = (((offset_ & 0x7800) | 0x400) >> 7) |
+                           (*((int*)packet_data_) & 0xFFFEFF08);
+  offset_ += XMAContextData::kBytesPerBlock; // Sequence number
+
+  packet_->data = packet_data_;
+  packet_->size = XMAContextData::kBytesPerBlock;
+
+  // Re-initialize the context with new sample rate and channels
+  if (context_->sample_rate != sample_rate || context_->channels != channels) {
+    context_->sample_rate = sample_rate;
+    context_->channels = channels;
+
+    // We have to reopen the codec so it'll realloc whatever data it needs.
+    // TODO: Find a better way.
+    avcodec_close(context_);
```

Comment:
If this ends up being super expensive we can pool the AudioDecoders - will have to watch it